### PR TITLE
BUG: groupby.idxmax/idxmin consistently raise on unobserved categorical

### PR DESCRIPTION
**Pull request by rhshadrach**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Part of #10694, but doesn't close it fully.

 - When given a non-ordered category with unobserved categories, we currently raise about the unobserved categories. We should instead raise about the non-orderedness for consistency with min/max.
 - _python_apply_general does not return the correct dtype when there is a CategoricalIndex will NaN values.
 - _python_apply_general fails on a single grouping with unobserved categories even when we're call it from transform where the unobserved categories should have no impact.
 - _python_apply_general with an empty DataFrame with no numeric columns returns all the columns even when `numeric_only=True`.

All of these are fixed in #54234

### CODE REVIEW

**Review by mroeschke — APPROVED** [review]
LGTM

**Comment by rhshadrach** [comment]
Friendly ping @WillAyd @jbrockmendel

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -2015,10 +2015,16 @@ def _transform(self, func, *args, engine=None, engine_kwargs=None, **kwargs):
             with com.temp_setattr(self, "as_index", True):
                 # GH#49834 - result needs groups in the index for
                 # _wrap_transform_fast_result
-                if engine is not None:
-                    kwargs["engine"] = engine
-                    kwargs["engine_kwargs"] = engine_kwargs
-                result = getattr(self, func)(*args, **kwargs)
+                if func in ["idxmin", "idxmax"]:
```
Comment (WillAyd) [inline pandas/core/groupby/groupby.py:2018]: Is it not possible to stick with the same engine pattern that is in place for this? At first glance I'm wondering what makes these different from `min`/`max`` that requires branching here

File: pandas/core/groupby/groupby.py [hunk 2]
```diff
@@ -5720,6 +5726,105 @@ def sample(
...
+                result_len = len(self.grouper.result_index)
+            assert result_len <= expected_len
+            has_unobserved = result_len < expected_len
+
+            raise_err: bool | np.bool_ = not ignore_unobserved and has_unobserved
+            # Only raise an error if there are columns to compute; otherwise we return
+            # an empty DataFrame with an index (possibly including unobserved) but no
+            # columns
+            data = self._obj_with_exclusions
+            if isinstance(data, DataFrame):
```
Comment (WillAyd) [inline pandas/core/groupby/groupby.py:5779]:
```suggestion
            if raise_err and isinstance(data, DataFrame):
```

I don't think we even need to go down this path if `raise_err` isn't True to being with right?

File: pandas/core/groupby/groupby.py [hunk 3]
```diff
@@ -5720,6 +5726,105 @@ def sample(
...
+                if numeric_only:
+                    data = data._get_numeric_data()
+                raise_err &= len(data.columns) > 0
```
Comment (WillAyd) [inline pandas/core/groupby/groupby.py:5782]: If the above is correct this can become simple assignment

File: pandas/core/generic.py [hunk 4]
```diff
@@ -11895,7 +11895,7 @@ def _logical_func(
 
     def any(
         self,
-        axis: Axis = 0,
+        axis: Axis | None = 0,
```
Comment (jbrockmendel) [inline pandas/core/generic.py:11898]: unrelated?

File: .github/workflows/code-checks.yml [hunk 5]
```diff
@@ -124,7 +124,7 @@ jobs:
       run: |
         cd asv_bench
         asv machine --yes
-        asv run --quick --dry-run --durations=30 --python=same
+        asv run --quick --dry-run --durations=30 --python=same --show-stderr
```
Comment (jbrockmendel) [inline .github/workflows/code-checks.yml:127]: is this something we want to keep or just for debugging this PR?

File: pandas/core/groupby/groupby.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/groupby/groupby.py:2018]: i think this is the same issue as i asked about somewhere else: _idxmax_idxmin accepts ignore_unobserved while idxmin/idxmax do not

Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:2018]: The main difference in behavior is that min/max do not raise on unobserved categories, while idxmin and idxmax do.

<details>
<summary>Example - agg</summary>

```
df = pd.DataFrame({'a': pd.Categorical([1, 1, 2], categories=[1, 2, 3]), 'b': [3, 4, 5]})
gb = df.groupby('a', observed=False)
result = gb.min()
print(result)
#      b
# a     
# 1  3.0
# 2  5.0
# 3  NaN
```

</details>

However, the fact that we don't do something special for min/max means that transform unnecessarily coerces to float:

<details>
<summary>Example - transform</summary>

```
df = pd.DataFrame({'a': pd.Categorical([1, 1, 2], categories=[1, 2, 3]), 'b': [3, 4, 5]})
gb = df.groupby('a', observed=False)
result = gb.transform('min')
print(result)
#      b
# 0  3.0
# 1  3.0
# 2  5.0
```

</details>

I consider this a bug in min/max with transform.

File: pandas/core/groupby/groupby.py (hunk 2 above)
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:5779]: Yea - `not raise_err`, right?

File: pandas/core/groupby/groupby.py (hunk 1 above)
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:2018]: I've opened #55326 to track

File: pandas/core/generic.py (hunk 4 above)
Comment (rhshadrach) [reply pandas/core/generic.py:11898]: mypy discovered this via my use of `.any(None)` in `_idxmax_idxmin` here.

File: .github/workflows/code-checks.yml (hunk 5 above)
Comment (rhshadrach) [reply .github/workflows/code-checks.yml:127]: I think it'd be useful to keep. We run the ASVs on the CI just to see if they error out, so it seems to make sense to also say what the error was.

File: pandas/core/groupby/groupby.py (hunk 1 above)
Comment (WillAyd) [reply pandas/core/groupby/groupby.py:2018]: Ah OK thanks - that is helpful context. So do you see the solution for min/max being the same as what you have for idxmin/idxmax here? 

I think the broader issue is that we've wanted over time to move away from branching for function specializations within groupby. If that still holds true then I wonder what prevents us from sticking with the existing kwargs interface to solve both this PR and eventually solve min/max's issue

Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:2018]: I'd like to explore different solutions for min/max and other aggregations, but I don't know what that could be at this time.

I don't understand what you're suggesting with kwargs.

Comment (WillAyd) [reply pandas/core/groupby/groupby.py:2018]: Sorry I thought the engine / engine_kwargs were specialized arguments for each function implementation, but I see now those are meant for numba.

The numba functions are UDFs right? I'm assuming from the branch here that we would never want to pass numba arguments to `_idxmax_idxmin`

Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:2018]: Correct - we never get here when using numba.

**Comment by rhshadrach** [comment]
@WillAyd - are you good here?

File: pandas/core/groupby/groupby.py [hunk 6]
```diff
@@ -2015,10 +2015,16 @@ def _transform(self, func, *args, engine=None, engine_kwargs=None, **kwargs):
...
                 # GH#49834 - result needs groups in the index for
                 # _wrap_transform_fast_result
-                if engine is not None:
-                    kwargs["engine"] = engine
-                    kwargs["engine_kwargs"] = engine_kwargs
-                result = getattr(self, func)(*args, **kwargs)
+                if func in ["idxmin", "idxmax"]:
+                    func = cast(Literal["idxmin", "idxmax"], func)
+                    result = self._idxmax_idxmin(
+                        func, ignore_unobserved=True, axis=self.axis, **kwargs
```
Comment (WillAyd) [inline pandas/core/groupby/groupby.py:2021]: We definitely don't need *args here right? Seems like something could get discarded compared to the other branch

**Review by WillAyd — COMMENTED** [review]
minor comment/question but otherwise I think this looks good

File: pandas/core/groupby/groupby.py (hunk 6 above)
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py]: Ack - thanks! Fixed and test added.

**Review by WillAyd — APPROVED** [review]
Lgtm

File: pandas/core/groupby/groupby.py [hunk 7]
```diff
@@ -5720,6 +5724,113 @@ def sample(
...
+        ignore_unobserved: bool = False,
+        axis: Axis | None | lib.NoDefault = lib.no_default,
+        skipna: bool = True,
+        numeric_only: bool = False,
+    ):
+        """Compute idxmax/idxmin.
+
+        Parameters
+        ----------
+        how: {"idxmin", "idxmax"}
```
Comment (jbrockmendel) [inline pandas/core/groupby/groupby.py:5739]: nitpick i think missing space between "how" and colon

Comment (jbrockmendel) [reply pandas/core/groupby/groupby.py:5739]: also usually but not always we have single-quotes inside these (no idea why and i genuinely dont care, extra since this is private)

File: pandas/core/groupby/groupby.py [hunk 8]
```diff
@@ -5720,6 +5724,113 @@ def sample(
...
+                    method = getattr(df, how)
+                    return method(axis=axis, skipna=skipna, numeric_only=numeric_only)
+
+                func.__name__ = how
+                result = self._python_apply_general(
+                    func, self._obj_with_exclusions, not_indexed_same=True
+                )
+        except ValueError as err:
+            name = "argmax" if how == "idxmax" else "argmin"
+            if f"attempt to get {name} of an empty sequence" in str(err):
```
Comment (jbrockmendel) [inline pandas/core/groupby/groupby.py:5812]: would this be simpler if we just said changed "arg" to "idx" in the cython method with a comment as to why we are using an apparently-wrong message?

File: pandas/core/groupby/groupby.py (hunk 7 above)
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:5739]: Thanks - fixed in #54234

File: pandas/core/groupby/groupby.py (hunk 8 above)
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:5812]: This arises from a call to NumPy's argmin in nanops.nanargmin:

    pd.Series().idxmin()
    # ValueError: attempt to get argmin of an empty sequence

In #54234, this remains only for the `axis=1` case, and so once that's deprecated, this code will be removed entirely.
